### PR TITLE
Fixed #20147 -- Added HttpRequest.headers.

### PR DESCRIPTION
**Pull request by santiagobasulto**
Original PR (with discussion included): https://github.com/django/django/pull/9925

I've rebased from master and squashed all the commits into one.

### CODE REVIEW

**Comment by santiagobasulto** [comment]
👋 folks, there's only one build failing with: `ModuleNotFoundError: No module named 'memcache'`
I haven't changed any dependency or setup.py file, do you have an idea of what might be going on?

Thanks!

File: docs/releases/2.1.txt [hunk 1]
```diff
@@ -211,6 +211,17 @@ Requests and Responses
   sets the ``Content-Disposition`` header to make the browser ask if the user
   wants to download the file. ``FileResponse`` also tries to set the
   ``Content-Type`` and ``Content-Length`` headers where appropriate.
+* Added :attr:`.HttpRequest.headers`.
```
Comment (auvipy) [inline docs/releases/2.1.txt]: shouldn't this moved to 2.2.txt?

Comment (auvipy) [reply docs/releases/2.1.txt]: you can drop this

**Comment by santiagobasulto** [comment]
There seems to be an error with jenkins not finding the right commit after the rebase. Anybody has an idea how I could fix this?

**Comment by auvipy** [comment]
everything seems passing

**Comment by santiagobasulto** [comment]
You're right, when I submitted the PR originally, it was failing, it's working now 👍

**Comment by santiagobasulto** [comment]
@carltongibson @timgraham PR finally rebased and working. Thanks for all your support. Gonna say this once again, this community is amazing.

File: django/utils/datastructures.py [hunk 2]
```diff
@@ -280,3 +281,84 @@ def __getitem__(self, key):
...
+            raise ValueError("dictionary update sequence element #{} has "
+                             "length {}; 2 is required".format(i, len(elem)))
+
+        if not isinstance(elem[0], str):
+            raise ValueError('Element key invalid, only strings are allowed')
+
+        yield tuple(elem)
+
+
+class ImmutableCaseInsensitiveDict(Mapping):
```
Comment (jschneier) [inline django/utils/datastructures.py]: This isn't actually Immutable. I didn't trace far back enough to see if the WSGI `environ` is immutable but cursory inspection would indicate that `.META` is not actually immutable although it would probably be a bad idea to manipulate it (I might be wrong on this point).

File: django/utils/datastructures.py [hunk 3]
```diff
@@ -280,3 +281,84 @@ def __getitem__(self, key):
...
+        return (original_key for original_key, value in self._store.values())
+
+    def __repr__(self):
+        return repr({key: value for key, value in self._store.values()})
+
+    def copy(self):
+        return self
+
+
+class EnvironHeaders(ImmutableCaseInsensitiveDict):
```
Comment (adamchainz) [inline django/utils/datastructures.py]: this class has a bunch of details of WSGI environ so I think it more accurately belongs in `django.http.request`

**Comment by santiagobasulto** [comment]
Hello @carltongibson @timgraham 👋 ! You think I could merge this one or should I make any other improvement?

**Comment by carltongibson** [comment]
Hi @santiagobasulto. As far as I can see the comments from @jschneier and @adamchainz remain unaddressed. 

Just dropping `Immutable` from the naming may be enough to address @jschneier's point (???)

Overall it seems quite good to me. If you can address the points and then uncheck `Patch needs improvement` on the ticket, I'll have another look. (Seems like a nice feature for the amount of code.)

**Comment by santiagobasulto** [comment]
Great, thanks @carltongibson. I'll get on it and update soon.

File: docs/ref/request-response.txt [hunk 4]
```diff
@@ -135,7 +135,9 @@ All attributes should be considered read-only, unless stated otherwise.
 
 .. attribute:: HttpRequest.META
 
-    A dictionary containing all available HTTP headers. Available headers
+    A dictionary containing the `WSGI environ \
```
Comment (timgraham) [inline docs/ref/request-response.txt]: This change looks unrelated to the patch and should be submitted separately.

File: docs/ref/request-response.txt [hunk 5]
```diff
@@ -160,13 +162,38 @@ All attributes should be considered read-only, unless stated otherwise.
...
     ``HTTP_X_BENDER``.
+    See :attr:`~HttpRequest.headers` for more convenient header access (``request.headers['User-Agent']``).
 
     Note that :djadmin:`runserver` strips all headers with underscores in the
     name, so you won't see them in ``META``. This prevents header-spoofing
     based on ambiguity between underscores and dashes both being normalizing to
     underscores in WSGI environment variables. It matches the behavior of
     Web servers like Nginx and Apache 2.4+.
 
+.. _headers:
```
Comment (timgraham) [inline docs/ref/request-response.txt]: This line isn't required.

File: docs/ref/request-response.txt [hunk 6]
```diff
@@ -160,13 +162,38 @@ All attributes should be considered read-only, unless stated otherwise.
...
+    .. versionadded:: 2.2
+
+    A case-insensitive dictionary containing the headers passed by the client.
+    The name of each header is Title-Cased (``User-Agent``) when
+    headers are printed or inspected (``str``, ``repr``).
+
+    When accessing elements with ``headers.get``, ``headers[]`` or
+    checking membership (``in``), case is ignored. Examples::
+
+            'User-Agent' in request.headers  # True
```
Comment (timgraham) [inline docs/ref/request-response.txt]: This block is indent 4 spaces too much.

File: tests/requests/tests.py [hunk 7]
```diff
@@ -806,3 +806,121 @@ def test_request_path_begins_with_two_slashes(self):
         for location, expected_url in tests:
             with self.subTest(location=location):
                 self.assertEqual(request.build_absolute_uri(location=location), expected_url)
+
+
+class CaseInsensitiveHeadersTestCase(SimpleTestCase):
```
Comment (timgraham) [inline tests/requests/tests.py]: TestCase -> Tests

File: docs/ref/request-response.txt [hunk 8]
```diff
@@ -135,7 +135,9 @@ All attributes should be considered read-only, unless stated otherwise.
 
 .. attribute:: HttpRequest.META
 
-    A dictionary containing all available HTTP headers. Available headers
+    A dictionary containing the `WSGI environ \
+    <https://www.python.org/dev/peps/pep-0333/#environ-variables>`_, including all available HTTP headers.
```
Comment (timgraham) [inline docs/ref/request-response.txt]: Documentation should be wrapped at 79 characters.

File: tests/utils_tests/test_datastructures.py [hunk 9]
```diff
@@ -155,3 +155,247 @@ def f(x):
...
+
+class ImmutableCaseInsensitiveDictTests(SimpleTestCase):
+    def setUp(self):
+        self.dict_1 = ImmutableCaseInsensitiveDict({
+            'Accept': 'application/json',
+            'content-type': 'text/html'
+        })
+
+    def test_create_dict_with_invalid_values(self):
+        with self.assertRaises(ValueError):
```
Comment (timgraham) [inline tests/utils_tests/test_datastructures.py]: Use `assertRaisesMessage()`.

File: tests/utils_tests/test_datastructures.py [hunk 10]
```diff
@@ -155,3 +155,247 @@ def f(x):
...
+            ImmutableCaseInsensitiveDict([('Key1', 'Val1'), 'Key2'])
+
+    def test_list(self):
+        self.assertEqual(
+            sorted(list(self.dict_1)),
+            sorted(['Accept', 'content-type']))
```
Comment (timgraham) [inline tests/utils_tests/test_datastructures.py]: trailing ) goes on the next line (check your entire patch for this style)

File: tests/utils_tests/test_datastructures.py [hunk 11]
```diff
@@ -155,3 +155,247 @@ def f(x):
...
+
+        self.assertEqual(str(dict_1), str({
+            'Accept': 'application/json',
+        }))
+        self.assertEqual(str(dict_2), str({
+            'content-type': 'text/html'
+        }))
+
+    def test_equals(self):
+        self.assertTrue(self.dict_1 == {
```
Comment (timgraham) [inline tests/utils_tests/test_datastructures.py]: Why not assertEqual?

File: tests/utils_tests/test_datastructures.py [hunk 12]
```diff
@@ -155,3 +155,247 @@ def f(x):
...
+        self.assertEqual(self.dict_1['Accept'], 'application/json')
+        self.assertEqual(self.dict_1['accept'], 'application/json')
+        self.assertEqual(self.dict_1['aCCept'], 'application/json')
+
+        self.assertEqual(self.dict_1['content-type'], 'text/html')
+        self.assertEqual(self.dict_1['Content-Type'], 'text/html')
+        self.assertEqual(self.dict_1['Content-type'], 'text/html')
+
+    def test_membership(self):
+        self.assertTrue('Accept' in self.dict_1)
```
Comment (timgraham) [inline tests/utils_tests/test_datastructures.py]: assertIn

File: tests/utils_tests/test_datastructures.py [hunk 13]
```diff
@@ -155,3 +155,247 @@ def f(x):
...
+            'Cookie': '%7B%22hello%22%3A%22world%22%7D;another=value'})
+
+    def test_custom_and_special_services_headers(self):
+        environ = {
+            'CONTENT_TYPE': 'text/html',
+            'CONTENT_LENGTH': '100',
+            'HTTP_ACCEPT': '*',
+            'HTTP_HOST': 'example.com',
+            'HTTP_USER_AGENT': 'python-requests/1.2.0',
+
```
Comment (timgraham) [inline tests/utils_tests/test_datastructures.py]: blank line not needed

File: docs/ref/request-response.txt [hunk 14]
```diff
@@ -160,13 +162,38 @@ All attributes should be considered read-only, unless stated otherwise.
...
+            'user-agent' in request.headers  # True
+
+            # getitem:
+            request.headers['User-Agent']  # Returns the value of User-Agent
+            request.headers['user-agent']  # Returns the value of User-Agent
+
+            # .get():
+            request.headers.get('User-Agent')  # Returns the value of User-Agent
+            request.headers.get('user-agent')  # Returns the value of User-Agent
+
```
Comment (timgraham) [inline docs/ref/request-response.txt]: one blank line please

**Comment by carltongibson** [comment]
Hi @santiagobasulto. 

This still looks good. It would be a super addition for v2.2. Are you up for pushing it over the finish line? (If not we'll see if we can pick it up...) 

Re @jschneier's query on immutability, from [PEP 3333](https://www.python.org/dev/peps/pep-3333/#specification-details) (the WSGI spec):  

> The `environ` parameter is a dictionary object, containing CGI-style environment variables. This object **must** be a builtin Python dictionary (not a subclass, UserDict or other dictionary emulation), and the application is allowed to modify the dictionary in any way it desires.

So it's not immutable by design. 

Your `ImmutableCaseInsensitiveDict` implements `Mapping` (which is read-only, vs `MutableMapping`) so maybe just `CaseInsensitiveMapping`, not worrying about actual immutability. 
`Request.headers` is then (via your `EnvironHeaders`) a case-insensitive read-only accessor. (Just what we want.)

(Anybody wanting to actually mutate `Request.META` — Why? 🙂 — can access it directly.) 

Moving `EnvironHeaders` as per @adamchainz's review and making the adjustments Tim suggested shouldn't leaving it far off.

**Comment by santiagobasulto** [comment]
Hey @carltongibson! Yes, I'll get to this immediately and push the fixes by tomorrow. Sorry, I've been quite busy lately but I don't want to keep postponing this.

File: tests/requests/tests.py [hunk 15]
```diff
@@ -806,3 +806,121 @@ def test_request_path_begins_with_two_slashes(self):
...
+    }
+
+    def test_base_request_headers(self):
+        request = HttpRequest()
+        request.META = self.TESTING_ENVIRON
+
+        self.assertEqual(dict(request.headers), {
+            'Content-Type': 'text/html',
+            'Content-Length': '100',
+            'Accept': '*', 'Host': 'example.com',
```
Comment (zachborboa) [inline tests/requests/tests.py]:
```suggestion
            'Accept': '*',
            'Host': 'example.com',
```

File: tests/requests/tests.py [hunk 16]
```diff
@@ -806,3 +806,121 @@ def test_request_path_begins_with_two_slashes(self):
...
+            'X-Custom-Header-1': 'custom_header_1',
+            'X-Custom-Header-2': 'custom_header_2'})
+
+    def test_wsgi_request_headers(self):
+        request = WSGIRequest(self.TESTING_ENVIRON)
+
+        self.assertEqual(dict(request.headers), {
+            'Content-Type': 'text/html',
+            'Content-Length': '100',
+            'Accept': '*', 'Host': 'example.com',
```
Comment (zachborboa) [inline tests/requests/tests.py]:
```suggestion
            'Accept': '*',
            'Host': 'example.com',
```

File: tests/requests/tests.py (hunk 15 above)
Comment (santiagobasulto) [reply tests/requests/tests.py]: Thanks @zachborboa!

**Comment by carltongibson** [comment]
Hi Santiago. Thanks for the work! I will review on Tuesday, so no panic. 🙂

File: django/utils/datastructures.py (hunk 2 above)
Comment (carltongibson) [reply django/utils/datastructures.py]: @santiagobasulto What are your thoughts on renaming this to `CaseInsensitiveMapping`? `Mapping` being the stdlib ABC you're implementing. (c.f. `ImmutableList.complain()`: we're taking measures there to avoid mutation. I don't think we want or need to get into any of that here.)

**Comment by santiagobasulto** [comment]
@carltongibson Agree completely on the Mapping suggestion, I'll fix it now. I somehow missed one of Tim fixes. I'll update everything together.

**Comment by maguayo** [comment]
Looking good! I love the changes

File: docs/ref/request-response.txt (hunk 4 above)
Comment (santiagobasulto) [reply docs/ref/request-response.txt]: I think that removing this and submitting it separately would lack the context later; what I did was just improving the docs a little bit more. From my experience, newcomers feel extremely confused when they find the prefix `HTTP_`.

File: docs/ref/request-response.txt (hunk 14 above)
Comment (santiagobasulto) [reply docs/ref/request-response.txt]: There's only one line here Tim. You mean remove all the lines altogether?

File: docs/ref/request-response.txt (hunk 4 above)
Comment (carltongibson) [reply docs/ref/request-response.txt]: Just the PR is fine. 🙂

**Comment by santiagobasulto** [comment]
@carltongibson all done now, I'll submit the doc changes in a separate PR. I'll take some time to read the whole text and see if there are other improvements to do. Thank so much again for your patience.

**Comment by santiagobasulto** [comment]
@carltongibson @timgraham hey folks 👋! Is there anything else you'd like me to change here? I 've addressed all your comments and seems like everything is passing. January 14th is the deadline for 2.2 features, right?

**Comment by timgraham** [comment]
Don't forget to uncheck "Patch needs improvement" on the ticket so that it appears in the review queue.

**Comment by santiagobasulto** [comment]
@timgraham I forgot about that 🤦‍♂️, thanks for the help.

**Comment by collinanderson** [comment]
@santiagobasulto If you get a chance, do you want to update the doc string for CaseInsensitiveMapping with the text that carltongibson suggested?

> Mapping allowing case-insensitive key lookups. Original case of keys is preserved for iteration and string representation.

Otherwise, I like this PR. Thanks so much! In same ways it would be nice if you changed one dict it would change the other, but I don't think that's going to be an issue.

**Comment by collinanderson** [comment]
@santiagobasulto yeah I don't think updating headers is worth it.

File: django/utils/datastructures.py [hunk 17]
```diff
@@ -280,3 +281,63 @@ def __getitem__(self, key):
         if use_func:
             return self.func(value)
         return value
+
+
+def _destruct_iterable_mapping_values(data):
+    for i, elem in enumerate(data):
+        if len(elem) != 2:
+            raise ValueError("dictionary update sequence element #{} has "
```
Comment (timgraham) [inline django/utils/datastructures.py]: Use hanging indent as described in [Python coding style](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/coding-style/#python-style).

File: django/http/request.py [hunk 18]
```diff
@@ -572,3 +578,38 @@ def validate_host(host, allowed_hosts):
     Return ``True`` for a valid host, ``False`` otherwise.
     """
     return any(pattern == '*' or is_same_domain(host, pattern) for pattern in allowed_hosts)
+
+
+class EnvironHeaders(CaseInsensitiveMapping):
+    IGNORE_EXCEPTIONS = {'HTTP_CONTENT_TYPE', 'HTTP_CONTENT_LENGTH'}
```
Comment (timgraham) [inline django/http/request.py]: I think this deserves a comment. Why are these headers "ignored" (whatever that means). Why are the headers below "special unchanged" (whatever that means).

File: django/utils/datastructures.py [hunk 19]
```diff
@@ -280,3 +281,63 @@ def __getitem__(self, key):
...
+
+    def __getitem__(self, key):
+        return self._store[key.lower()][1]
+
+    def __len__(self):
+        return len(self._store)
+
+    def __eq__(self, other):
+        if not isinstance(other, Mapping):
+            return NotImplemented
```
Comment (timgraham) [inline django/utils/datastructures.py]: I'd expect to see a test for this. Please audit test coverage carefully and make sure all lines are covered.

File: tests/utils_tests/test_datastructures.py [hunk 20]
```diff
@@ -148,3 +148,121 @@ def f(x):
...
+        )
+
+    def test_dict(self):
+        self.assertEqual(dict(self.dict_1), {
+            'Accept': 'application/json',
+            'content-type': 'text/html'
+        })
+
+    def test_repr(self):
+        dict_1 = CaseInsensitiveMapping({
```
Comment (timgraham) [inline tests/utils_tests/test_datastructures.py]: Use a single line throughout where possible.

File: tests/utils_tests/test_datastructures.py [hunk 21]
```diff
@@ -148,3 +148,121 @@ def f(x):
...
+        self.assertIn('accept', self.dict_1)
+        self.assertIn('aCCept', self.dict_1)
+
+        self.assertIn('content-type', self.dict_1)
+        self.assertIn('Content-Type', self.dict_1)
+
+    def test_delitem(self):
+        """del should raise a TypeError because this is immutable"""
+        # Preconditions
+        self.assertTrue('Accept' in self.dict_1)
```
Comment (timgraham) [inline tests/utils_tests/test_datastructures.py]: assertIn

File: django/http/request.py (hunk 18 above)
Comment (collinanderson) [reply django/http/request.py]: Maybe just reference that these don't get the prefix as specified in https://www.python.org/dev/peps/pep-0333/

I understand why `SPECIAL_UNCHANGED_HEADERS` is needed but I don't understand why `IGNORE_EXCEPTIONS` is needed. Do we ever get incoming `HTTP_CONTENT_TYPE` keys?

Comment (santiagobasulto) [reply django/http/request.py]: I remember I did some research about the possible variations of multiple WSGI servers and some users noticed that some servers were not honoring PEP-333 and included `HTTP_CONTENT_TYPE`. 

I found that Python's [`wsgiref.validate`](https://docs.python.org/3.6/library/wsgiref.html) module [raises an exception](https://github.com/python/cpython/blob/master/Lib/wsgiref/validate.py#L310) if ever found.

Options:

a) leave it as it is: A robust alternative. If someone is using a poorly designed wsgi server, it'll still work. Maybe issue a warning.
b) Remove it all together. Assume that wsgi implementations will honor PEP 333.
c) Make it fail if ever found one (raise an exception). Similar to what `wsgiref.validate` is doing. This is kind of redundant, i don't like it TBH.

File: docs/ref/request-response.txt [hunk 22]
```diff
@@ -160,13 +160,37 @@ All attributes should be considered read-only, unless stated otherwise.
     underscores and adding an ``HTTP_`` prefix to the name. So, for example, a
     header called ``X-Bender`` would be mapped to the ``META`` key
     ``HTTP_X_BENDER``.
+    See :attr:`~HttpRequest.headers` for more convenient header
```
[the remark this answers is not in the document]
Comment (santiagobasulto) [reply docs/ref/request-response.txt]: Does "more intuitive" sound nicer? There's an example after it `request.headers['User-Agent']`, which makes it more obvious, or adds extra context.

**Comment by santiagobasulto** [comment]
Hey folks 👋 , just finished addressing comments by @timgraham; also improved docs with the suggestions by @carltongibson (replace "dict" with case insensitive mapping).
The only open thing to resolve is in this comment: https://github.com/django/django/pull/10171#discussion_r246975198

I have the whole day free tomorrow (traveling to Raleigh NC for the first time, if you have recommendations of cool things to do they're welcome 😁). So I'd like to finish every open issue on this one and get it merged between the feature freeze. IMO having `headers` will really help beginners.

File: django/http/request.py (hunk 18 above)
Comment (collinanderson) [reply django/http/request.py]: Ok, yeah it looks like uwsgi might have, at least at some point, both CONTENT_TYPE and HTTP_CONTENT_TYPE.  So, yeah, I think ignoring HTTP_CONTENT_TYPE and trusting only CONTENT_TYPE is a good idea (a).

Comment (santiagobasulto) [reply django/http/request.py]: @timgraham all points of your review addressed and I added the comments on these variables, just in case.

**Comment by chicheng** [comment]
@santiagobasulto Could you document what will be returned if duplicated header present in the request?

https://tools.ietf.org/html/rfc7230#section-3.2.2
>    A sender MUST NOT generate multiple header fields with the same field name in a message unless either the entire field value for that header field is defined as a comma-separated list [i.e., #(values)] or the header field is a well-known exception (as noted below).

File: django/http/request.py (hunk 18 above)
Comment (timgraham) [reply django/http/request.py]: I don't think Django has existing workarounds for this issue, e.g. https://github.com/django/django/blob/f021c110d02fd7ca32ae56f511b46e5d138b6c73/django/core/handlers/wsgi.py#L83
I think it adds necessary complexity.

File: django/http/request.py [hunk 23]
```diff
@@ -572,3 +578,43 @@ def validate_host(host, allowed_hosts):
...
+        ) for header_name, value in environ.items())
+        headers = {
+            header: value for header, value in header_name_generator if header
+        }
+
+        super().__init__(headers)
+
+    @classmethod
+    def _style_header_name(cls, header_name):
+        return header_name.title()
```
Comment (timgraham) [inline django/http/request.py]: Does this need to be a separate method? Seems unnecessary to me.

File: tests/requests/tests.py [hunk 24]
```diff
@@ -830,3 +830,251 @@ def test_request_path_begins_with_two_slashes(self):
...
+        self.assertEqual(request.headers.get('Pragma'), 'no-cache')
+        self.assertEqual(request.headers.get('Content-Type'), 'text/html')
+        self.assertEqual(request.headers.get('Content-Length'), '100')
+
+
+class EnvironHeadersTests(SimpleTestCase):
+    def test_parse_cgi_headers_basics(self):
+        parser = EnvironHeaders.parse_cgi_header_name
+
+        self.assertEqual(parser('HTTP_ACCEPT'), 'Accept')
```
Comment (timgraham) [inline tests/requests/tests.py]: Similar to my "use a simpler dictionary" comment above, this fells like overkill. The logic is fairly simple. What does "'HTTP_X_CUSTOM_HEADER_1" test that "HTTP_X_CUSTOM_HEADER_2" doesn't? Also you can use a loop with `subTest()` to simplify rather than have separate methods. For example:
https://github.com/django/django/blob/f021c110d02fd7ca32ae56f511b46e5d138b6c73/tests/utils_tests/test_http.py#L156-L158

File: django/http/request.py (hunk 18 above)
Comment (santiagobasulto) [reply django/http/request.py]: @timgraham 👍

File: django/http/request.py (hunk 23 above)
Comment (santiagobasulto) [reply django/http/request.py]: This can be extended by subclasses and overwritten, that's why I defined it as a different method.

**Comment by adamchainz** [comment]
🎉
